### ISSUE TITLE
Allow closing timer immediately without waiting for close callback

### ISSUE THREAD

**Issue by ambrop72** [libuv/libuv#1536]
After uv_close is used on a timer, the application must wait for the close callback to be invoked before it is allowed to release the memory of the uv_timer_t. This means that for example the application may have to dynamically allocate memory for uv_timer_t so that its lifetime can be extended beyond the point when the application no longer needs the timer. If it was possible to close the timer immediately then uv_timer_t could be directly embedded into some object/class in the application avoiding additional memory allocation. This especially applies to C++ programs relying heavily on RAII.

My understanding is that this could easily be implemented. Specifically, in uv_close just after uv__timer_close, check if the callback is null and in that case call uv__finish_close and return (skipping uv__make_close_pending).

Such a change might break some existing applications; an application may be closing the timer with null callback but expecting that the timer is still alive (not closed) just after the call. I don't think any sane application would do that but if it is considered a problem, then instead of changing uv_close, a new function can be added for closing a timer without a callback (uv_timer_close?).

**Comment by txdv** [comment]
Interesting, all tests are green on my machine...

Looks like such a simple thing to do and it should actually save some overhead.

**Comment by saghul** [comment]
I don't think this is a good idea. While it might be possible for timers, it's not possible for other types of handles , thus creating an inconsistency in our API.

**Comment by txdv** [comment]
What handles exactly can't close immediate and why?

**Comment by saghul** [comment]
> What handles exactly can't close immediately and why?

Off the top of my head: signal handlers on Unix, and many more on Windows because we need to wait for IOCP cancellations.

This is by design, to accommodate for platform differences. I don't see this changing anytime soon, if ever.

> I guess the tests don't cover callback = null cases that is why everything is green currently.

Passing null as a callback is fine, as long as you never free the handle.

**Comment by ambrop72** [comment]
I agree it cannot work in general, but it should be possible for timers because timers just need to be removed from the heap and any other data structures (list of handles I think), there are no OS resources backing them (I believe this is the case in the Windows implementation too). In applications that use many timers this does matter.

What do you think about a separate uv_timer_close() for this purpose? It does not change the semantics of uv_close(), and applications could rely on the timer to be closed immediately if they call that instead of depending on which libuv version is used.

**Comment by saghul** [comment]
> I agree it cannot work in general, but it should be possible for timers because timers just need to be removed from the heap and any other data structures (list of handles I think), there are no OS resources backing them (I believe this is the case in the Windows implementation too). In applications that use many timers this does matter.

What if we want to change it in the future? Also, how about consistency? Right now `uv_close` works exactly the same for all types of handles. It's predictable and easy to understand. You suggest introducing an annomaly for what looks to me like a small convenience improvement.

**Comment by ambrop72** [comment]
I hope we agree that from the perspective of an application developer having an interface that allows immediate close is clearly better to one where you can't release the memory until you are told to. And for timers this is clearly possible to implement regardless of the platform. So I don't see why after adding this feature one would want to later remove it.

Why would applications care about consistency between timers and network connections and other resources? They have nothing in common except that both come from the same library, and I think that enforcing consistency is detrimental here.

Actually I think most of things that timers inherit from uv_handle does not really apply well to timers:
- uv_is_active tells you whether the timer is running. It is useful but overloaded, having an uv_timer_is_running would be more understandable and easier to find in the docs.
- uv_close: since the timer is constructed with a specific function there is no real benefit that you can close it with a common function, uv_timer_close() would be just as good for an application (and of course there is the complication this issue is about).
- uv_ref, uv_unref, uv_has_ref: don't see who would ever need this, applications can easily do their own reference counting if they need it.
- uv_handle_size: does seem to apply to timers indeed.

I'm not saying uv_timer_t should be turned around to not inherit uv_handle since that would break applications, but how can adding an uv_timer_close() possibly hurt?

**Comment by txdv** [comment]
> uv_ref, uv_unref, uv_has_ref: don't see who would ever need this, applications can easily do their own reference counting if they need it.

You want to create a timer which keeps the loop alive vs a timer that doesn't keep the loop alive? Seems like a reasonable thing to have for the user at his disposal.

**Comment by ambrop72** [comment]
> You want to create a timer which keeps the loop alive vs a timer that doesn't keep the loop alive?

I see, I suppose that could be useful to someone (but personally I think just uv_quit is good enough).

Concerning my pull request, I just taught I forgot `uv__handle_unref` in `uv_timer_stop_now`, but I see that it's redundant because the only thing it could do is decrement `active_handles`, but that would already have been done just earlier by `uv_timer_stop` if it was needed. Also, it may be nice to set the stopping and stopped flags, but I don't see how to do it nicely because the flags are different for unix and windows.

**Comment by moteus** [comment]
> I hope we agree that from the perspective of an application developer having an interface that allows immediate close is clearly better to one where you can't release the memory until you are told to.

If you use GC it is not as important.

>Why would applications care about consistency between timers and network connections and other resources? They have nothing in common except that both come from the same library, and I think that enforcing consistency is detrimental here.

For me they all are sources of events for event loop. So they all the same from this point of view.

>uv_close: since the timer is constructed with a specific function there is no real benefit that you can close it with a common function

Benefits in that you can travers via all handles and just close them in the same way without bothering about their types.

Do not forget that libuv loop may be closed and recreated in same process more than once.
So ability to close all handles in such cases is really important.

> uv_is_active tells you whether the timer is running. It is useful but overloaded, having an uv_timer_is_running would be more understandable and easier to find in the docs.

Again. It is useful to have same API for all handle types. It is allows use same code for common tasks.
E.g. in my Lua binding i can use `uv_is_active` to all handle to tell is it possible gc`ed them if there no 
other reference. E.g. if handle active it should be stay avaliable but if it is not active and there no other
reference it can be closed and cleared.
And this logic has no difference for timers or sockets or any other handle type.

**Comment by saghul** [comment]
> I hope we agree that from the perspective of an application developer having an interface that allows immediate close is clearly better to one where you can't release the memory until you are told to. And for timers this is clearly possible to implement regardless of the platform. So I don't see why after adding this feature one would want to later remove it.

You still don't seem to understand my concern about consistency.

> Why would applications care about consistency between timers and network connections and other resources? They have nothing in common except that both come from the same library, and I think that enforcing consistency is detrimental here.

I beg to differ.

> Actually I think most of things that timers inherit from uv_handle does not really apply well to timers:
>

You might want to consider having a look at the design: http://docs.libuv.org/en/v1.x/design.html We are not throwing code at a wall here, things have reasons.
 
> uv_is_active tells you whether the timer is running. It is useful but overloaded, having an uv_timer_is_running would be more understandable and easier to find in the docs.

Then we'd have 10 functions, one for each handle. No thanks.

> uv_close: since the timer is constructed with a specific function there is no real benefit that you can close it with a common function, uv_timer_close() would be just as good for an application (and of course there is the complication this issue is about).

I already explained this before.

> uv_ref, uv_unref, uv_has_ref: don't see who would ever need this, applications can easily do their own reference counting if they need it.

You don't seem to understand what those mean. They are not counters. Of course you can manage the lifetime of handles by hand, but it's a lot more of a hassle. As an example: a signal handler to signal SIGINT shouldn't keep the loop alive, so unref-ing it makes all the sense in the world.

**Comment by saghul** [comment]
@moteus Thanks, I couldn't have written it better.

**Comment by saghul** [comment]
@libuv/collaborators I move to close, this is all by design and don't see any strong enough argument to make the suggested changes. Let's have a quick vote.

-1.

**Comment by cjihrig** [comment]
-1

**Comment by ambrop72** [comment]
Please consider that the implementation itself is literally 6 lines of code. I don't disagree with the arguments presented by moteus. But still in C++ applications using RAII, it presents a problem and consistency is not helping. In such designs for example it doesn't make much sense to iterate through all handles to and close them because that would break abstractions everywhere (rather one would destruct objects that use libuv handles, taking the handles down with them).

**Comment by bnoordhuis** [comment]
I'm somewhat in favor, actually.  The ability to close-and-forget timers (or idle/prepare/check handles, and especially async handles) would make life easier for libuv users.

Anecdote time: in nearly every node.js pull request I've reviewed in the last year that deals with async handles, there has been at least one use-after-free bug where a handle's memory is freed before `uv_close()` is done with it.

(Yes, this issue is about timers but I'm unapologetically moving the goalposts here.)

**Comment by ambrop72** [comment]
If there is interest I can implement is for the other handle types that @bnoordhuis mentioned. From a quick look, this can indeed be done for all of those as there are no handle-specific resources that could take time to release, only data structures they need to be removed from.

**Comment by saghul** [comment]
If this is to happen, I'd like to see a [LEP](https://github.com/libuv/leps) written about it.

**Comment by txdv** [comment]
Basically it looks like handles without pending callbacks can be closed immediately.

It would be possible to close handles with pending requests/callbacks as well, if it is ok to call those callbacks not in the order that uv_run defines (call all write requests of a stream with ECANCELED immediately, not waiting for the appropriate place in uv_run) Is that a thing that you guys are willing to do?

**Comment by moteus** [comment]
Main question when it is safe free memory for this handle.
Now it is safe only after callback called.
Callback have to get valid handle in any case.

**Comment by ambrop72** [comment]
Here's a quick summary of what I found:
- Timers: do not have requests (req), they only need to be removed from the heap (and handle list and active handle refcount...). `uv__run_timers` is already safe against a timer that self-destructs or destructs another timer in a callback since timers are consumed directly from the top of the heap.
- Async: similar as with timers, they need to be removed from a list and doing so seems safe. See `uv_process_async_wakeup_req` (win) and `uv__async_io` (unix). In stable for windows a request is involved but I assume this feature is only considered for master.
- Prepare/check/idle: again same scenario, just need to remove it from the list and `uv__run_##name` seems safe in that event.

For the latter two (async, prepare/check/idle) they are dispatched by first moving the entire queue and then consuming that, but removing such a moved handle should still be safe. This can already happen, e.g. `uv__async_close`.

So for these I think that we can have a simple close function (perhaps `uv_close_now`) that guarantees the handle is closed immediately. I suggest that this feature is for now limited to that. I don't think it benefits applications much if they can "sometimes" close the handle immediately.

For others like tcp/udp on Windows there really is a complication since we need to wait for IO completions and buffers must remain available until then. Perhaps it would be possible to completely hide the async cleanup from the application. I'll think about that more...

**Comment by saghul** [comment]
Some random thought:  Instead of adding a new API, we could change the return value of `uv_close` to an `int`.  if no callback is passed and everything could be closed on the spot, it would return 0, if a callback is passed, or (even in the absence of a callback) the close cannot be done synchronously, it could return `UV_EINPROGRESS`.

But again, I want to see a LEP with a design, not scattered -around GitHub comments.

**Comment by ambrop72** [comment]
@txdv You mention pending callbacks but my understading is there isn't *generally* such a thing. If it is just about the handle being inserted into some list or other data structure, it can be removed at the time of the close, so long as there is no safety hazard in doing so (code that consumes the data structure needs to be written carefully to account callbacks removing handles from that data structure).

On Windows we have outstanding I/O operations which indeed need to complete before the handle (more specifically req) can be released. That could perhaps in be solved by libuv keeping alive only the resources actually needed for outstanding I/O operations to complete but it might need some redesign.

**Comment by ambrop72** [comment]
@saghul What is a LEP and how do I write one?

**Comment by txdv** [comment]
@ambrop72 Yeah, you are right, but there is a contract that the library is trying to enforce.

**Comment by saghul** [comment]
> What is a LEP and how do I write one?

@ambrop72 see https://github.com/libuv/leps

**Comment by saghul** [comment]
> It would be possible to close handles with pending requests/callbacks as well, if it is ok to call those callbacks not in the order that uv_run defines (call all write requests of a stream with ECANCELED immediately, not waiting for the appropriate place in uv_run) Is that a thing that you guys are willing to do?

@txdv I don't really know. Windows pipes are tricky, I don't think we can do that in all scenarios. Same goes for TCP connections were we emulate IOCP, I think.

**Comment by txdv** [comment]
I'm asking if it would be acceptable to move to calling the request callbacks on close rather within uv_run in the  next iteration (from a contract point of view) not if it is possible to code.

**Comment by saghul** [comment]
I don't know, we'd need to think about all the implications that might have.

**Comment by edwinchenloo** [comment]
i don't see the need to break api consistency.

When dealing with async events, one needs to be aware of lifetimes.  If the timer and event loop are on the stack they will be destroyed in reverse creation order so first create the loop then the timer.  If one wants to remove a timer but keep a loop alive that lives on a different function/stack, the timer was to be on the heap and the loop will let you know when to delete it.  It is a pattern consistent with all other handles.

It seems some users want to create a variable on the stack and not wait or block until the loop is finished with the object.  This defeats the purpose of having an async loop.  No matter what, one has to wait until the loop indicates the object is no longer needed (via the close callback).  You can then destroy your timer from the stack or delete from dynamic memory.

It is not that difficult to create an object that inherits or encapsulates a uv_timer_t and delete it on the close callback.

**Comment by cjihrig** [comment]
The associated LEP was closed a few months ago, and this has largely stalled out. I'm going to close this, but if someone wants to revisit https://github.com/libuv/leps/pull/11, feel free.
